brainiak/reconstruct/iem.py

```diff
@@ -0,0 +1,290 @@
+    regression (using fit()). Then, as part of predict(),
+    compute channel outputs for new functional data (done
+    by _predict_channel_responses()), and associate those
+    responses to a feature type. Score() computes a measure
+    of the error of the prediction based on known ground
+    truth.
+
+    This implementation assumes a circular (or half-
+    circular) feature domain. Future implementations might
+    generalize the feature input space, and increase the
+    possible dimensionality.
+
+    Parameters
+    ----------
+    n_channels: int, default 5, number of channels
+        The number of channels, or basis functions, to be used
+        in the inverted encoding model
+
+    range_start: double, default 0, beginning of range of
+        independent variable (usually degrees)
+
+    range_stop: double, default 180, end of range of
+        independent variable (usually degrees)
+
+
+    Attributes
+    ----------
+    C_: [n_channels, channel density] NumPy 2D array
+        matrix defining channel values
+
+    C_D_: [n_channels, channel density] NumPy 2D array
+        matrix defining channel independent variable (usually
+        in degrees)
+
+    W_: sklearn.linear_model model containing weight matrix that
+        relates estimated channel responses to response amplitude
+        data
+    """
+    def __init__(self, n_channels=5, range_start=0, range_stop=180):
+
+        self.n_channels = n_channels  # default = 5
+        self.range_start = range_start  # in degrees, 0 - 360, def=0
+        self.range_stop = range_stop  # in degrees, 0 - 360, def=180
+
+    def fit(self, X, y):
+        """Use data and feature variable labels to fit an IEM
+
+        Parameters
+        ----------
+        X: numpy matrix of voxel activation data. [observations, voxels]
+            Should contain the beta values for each observation or
+            trial and each voxel of training data.
+        y: numpy array of response variable. [observations]
+            Should contain the feature for each observation in X.
+
+        Returns
+        -------
+        iem: self.
+        """
+        # Check that there are channels specified
+        if self.n_channels < 2:
+            raise ValueError("Insufficient channels.")
+        # Check that there is enough data.. should be more
+        # samples than voxels (i.e. X should be tall)
+        shape_data = np.shape(X)
+        shape_labels = np.shape(y)
+        if len(shape_data) < 2:
+            raise ValueError("Not enough data")
+        else:
+            if np.size(X, 0) <= np.size(X, 1):
+                raise ValueError("Data Matrix ill-conditioned")
+            if shape_data[0] != shape_labels[0]:
+                raise ValueError(
+                    "Mismatched data samples and label samples")
+
+        self.C_, self.C_D_ = self._define_channels()
+        n_train = len(y)
+        F = np.empty((n_train, self.n_channels))
```

Comment:
I think this section can be benefited from additional comments. I guess this is trying to find the best channel for each training example to construct a design matrix?

Reply:
I added extra comments.